Mathlib/Tactic/Pushneg.lean

```diff
@@ -0,0 +1,119 @@
+
+def binderTelescope1 (e : Expr) (f : Expr → Expr → MetaM α) : MetaM α := do
+match e with
+| Expr.lam n t b d => withLocalDecl n d.binderInfo t fun x => f x (instantiate1 b x)
+| Expr.forallE n t b d => withLocalDecl n d.binderInfo t fun x => f x (instantiate1 b x)
+| _ => throwError "binder expected{indentExpr e}"
+
+partial def pushNegation (expr : Expr) : MetaM (Expr × Expr) := do
+match expr with
+| Expr.forallE n t b _ =>
+  binderTelescope1 expr fun x e => do
+    let (eNew, eqProof) ← pushNegation e
+    if eqProof.isAppOf ``Eq.refl
+    then return (expr, ← mkEqRefl expr)
+    else return (←mkForallFVars #[x] eNew, ← mkForallCongr (←mkLambdaFVars #[x] eqProof))
+| Expr.lam n t b _ =>
+  binderTelescope1 expr fun x e => do
+    let (eNew, eqProof) ← pushNegation e
+    if eqProof.isAppOf ``Eq.refl
+    then return (expr, ← mkEqRefl expr)
+    else return (←mkLambdaFVars #[x] eNew, ← mkFunExt (←mkLambdaFVars #[x] eqProof))
+| Expr.app e₁ e₂ _ =>
+   match expr.not? with
+   | some expr' =>
+      match expr'.getAppFnArgs with
+      | (`Not, #[e]) =>
+        let (eNew, eqProof) ← pushNegation e
+        let eqProof ← mkAppOptM ``not_not_eq #[none, none, eqProof]
+        return (eNew, eqProof)
+      | (`And, #[p, q]) =>
+        let (p, eqProof₁) ← pushNegation p
+        let (q, eqProof₂) ← pushNegation (mkNot q)
+        let eNew := mkForall `_ default p q
+        let eqProof ← mkAppOptM ``not_and_eq #[none, none,none, none, eqProof₁, eqProof₂]
+        return (eNew, eqProof)
+      | (`Or, #[p, q]) =>
+        let (p, eqProof₁) ← pushNegation (mkNot p)
+        let (q, eqProof₂) ← pushNegation (mkNot q)
+        let eNew := mkAnd p q
+        let eqProof ← mkAppOptM ``not_or_eq #[none, none,none, none, eqProof₁, eqProof₂]
+        return (eNew, eqProof)
+      | (`Exists, #[t, e]) =>
+        binderTelescope1 e fun x e => do
+          let (eNew, eqProof) ← pushNegation (mkNot e)
+          let eNew ← mkForallFVars #[x] eNew
+          let eqProof ← mkLambdaFVars #[x] eqProof
+          let eqProof ← mkAppOptM ``not_exists_eq #[none, none, none, eqProof]
+          return (eNew, eqProof)
+      | _ => match expr' with
+        | Expr.forallE _ t _ _ =>
+          binderTelescope1 expr' fun x e => do
+            let (eNew, eqProof) ← pushNegation (mkNot e)
+            let eNew ← mkLambdaFVars #[x] eNew
+            let level ← getLevel (←inferType eNew)
+            let eNew := mkAppN (mkConst ``Exists [level]) #[t, eNew]
+            let eqProof ← mkLambdaFVars #[x] eqProof
+            let eqProof ← mkAppOptM ``not_forall_eq #[none, none, none, eqProof]
+            return (eNew, eqProof)
+        | _ =>
+          let (eNew, eqProof) ← pushNegation expr'
+          let eqProof ← mkCongr (←mkEqRefl (mkConst `Not)) eqProof
+          return (mkNot eNew, eqProof)
+   | _ =>
+      let (e₁', eqProof₁) ← pushNegation e₁
+      let (e₂', eqProof₂) ← pushNegation e₂
+      let eqProof ← mkCongr eqProof₁ eqProof₂
```

Comment:
This will only work if e₁ has a non-dependent type.   I think it would be more robust to specifically check for Or/And (binary connectives) and Neg (unary connectives) and Exists here.

Reply:
Do you think you can exhibit a case where it doesn't work because e₁ is a dependent-type ?

Reply:
```lean
example (p : Bool) : decide (¬ ¬ p) = p := by push_neg
```